Improve handling of Spotify Top Tracks and compilations

cincodenada opened the pull request:
Currently, Spotify results are problematic for two reasons, which are both demonstrated in this screenshot of the results for a single New Pornographers album:
![2014-10-28-205030_331x546_scrot](https://cloud.githubusercontent.com/assets/479715/4820853/c4399478-5f1e-11e4-9953-abf5fc8be0a2.png)

Firstly, they are not assigned an album artist, so albums with different artists per track that should have a single album artist and therefore show up as one album when sorting by Album Artist are split up into two.  In this case, track 8 (Born With a Sound) features Amber Webber, so even though I am grouping by Album Artist, the album is split up between two artists.  When you add the album under The New Pornographers to your playlist, it is missing track 8, since that is filed under "The New Pornographers, Amber Webber".  My proposed solution here is to choose a "majority artist" by tallying up which artist shows up most often, and assigning that as the album artist.  This should be the "correct" solution in most cases, and even in edge cases (Broadway cast recordings, etc) will at least group the album into a single entry, albeit perhaps with an odd Album Artist entry.

The second issue is that the individual "top tracks" that Spotify provides retain their album title, so when sorted create one or more extra album entries, each of which only have a few songs in them, which is confusing, and not how Spotify presents them.  In the example above, all the top tracks are from one album, so we just end up with two copies of the album, one of which only has the 5 Top Tracks in it.

:Here's an example of the 5 Top Tracks being spread across multiple albums, demonstrating how things can get extra messy:

![2014-10-28-204628_357x793_scrot](https://cloud.githubusercontent.com/assets/479715/4820837/72df444c-5f1e-11e4-8a74-45bf86506dbc.png)

My proposed solution to this is to group these tracks into an artifical "Top Tracks" album, so that they show up in one place, and it is clear what they are.

This pull request implements both of those solutions, and the result looks like this:

![2014-10-28-204950_332x491_scrot](https://cloud.githubusercontent.com/assets/479715/4820873/1cbc0734-5f1f-11e4-8fd1-da6e76fd093d.png)

The duplicate album entries are gone, and the Top Tracks are clearly marked as such.

There are some other ways this could be done, but I implemented the most conservative changes as I saw them.  An option that is more invasive but perhaps more widely helpful would be instead of assigning the Majority Artist to the Album Artist, simply set each track's Artist to the Majority Artist, since the default sort options don't involve Album Artist.  This would be straightforward to implement, and show better results to more people, at the cost of losing accuracy of metadata.  The name for "Top Tracks" is also arbitrary, it could be "[Spotify] Top Tracks" to make clear it's a placeholder album, or whatever makes the most sense.

A note on the patch itself: I have successfully compiled against the 1.2.3 release, and these changes with that base are in the spotify-improvements-1.2 branch.  I have been unable to compile master starting at 56c9498, due to some apparent issues in my build environment with Gstreamer 1.x, but it's a small patch that doesn't touch any GStreamer things, so I can't imagine this patch against master will cause any problems.

cincodenada commented:
@ArnaudBienner, I arrived at a similar conclusion, and have implemented it as such, at least for the Spotify search.  "Top Tracks" results are now simply removed if they already exist in the album results (simple track/album comparison), which will take care of most of the user-facing issues, I believe.  Here's what the searches above now look like:
![2015-02-15-142829_346x357_scrot](https://cloud.githubusercontent.com/assets/479715/6205342/81b9c990-b51f-11e4-9c27-641221e2a7f9.png)
![2015-02-15-142913_314x579_scrot](https://cloud.githubusercontent.com/assets/479715/6205344/85f1df98-b51f-11e4-8642-d2a8f58f2658.png)

This also addresses the issue that @Asfaloth pointed out and the album name issue, since single tracks will still show up if they're not in the album results, and the "Top Tracks" fake album is done away with entirely.

I think given the constraints of the unified search interface, and the lack of info (as far as I could tell) from the Spotify search provider about what section the results were in, that this is probably the best overall solution - it's the one I'm using :)

cincodenada commented:
Is there anything further to be done on this PR?  I appreciate the feedback thus far - you all helped track down some failure cases that I hadn't run into.  I think my latest commit addresses all the concerns raised fairly well, and is an improvement over the current handling.

I was reminded to check in here, in fact, because I ran into an instance today that had exceptionally undesirable results: 
![2015-04-08-153511_404x656_scrot](https://cloud.githubusercontent.com/assets/479715/7056942/76909418-de05-11e4-8a18-6a3b08c6b06b.png)

This, of course, is just an extreme case of the album having many artists, without a unifying album artist, which is one of the two things this PR addresses.

amuttsch commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -1,4 +1,4 @@
-/* This file is part of Clementine.
+f/* This file is part of Clementine.
```
f?

amuttsch commented:
A wild `f` appread, please remove it, run `make format` and squash the commits. After this it can be merged.

cincodenada commented:
Indeed, will do.  Shall I squash the commits into one?

cincodenada commented:
Okay, possibly stupid question: how/from where do I run `make format`?  Is there a utility I need to install?  I remember running it before for other improvements, and I don't remember having to do anything fancy, but simply running `make format`, from the repo root, `bin`, or `bin/src` just gives me `*** No rule to make target 'format'.  Stop.`

I couldn't find any reference to it or instructions on the wiki either...I know I've done this before, but I can't figure out what I did!

amuttsch commented:
Normally from the bin dir. Have You merged Master recently? Since your pr is quite old, maybe make format wasn't added back then. 

On 15 April 2015 19:12:02 CEST, Joel Bradshaw <notifications@github.com> wrote:
>Okay, possibly stupid question: how/from where do I run `make format`? 
>Is there a utility I need to install?  I remember running it before for
>other improvements, and I don't remember having to do anything fancy,
>but simply running `make format`, from the repo root, `bin`, or
>`bin/src` just gives me `*** No rule to make target 'format'.  Stop.`
>
>I couldn't find any reference to it or instructions on the wiki
>either...I know I've done this before, but I can't figure out what I
>did!
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/clementine-player/Clementine/pull/4599#issuecomment-93489756

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

cincodenada commented:
Okay, turns out there were some missing dependencies for the new stuff that was preventing cmake from finishing.  Got that all straightened out, ran `make format` and squashed things down.  Should be good to go, I think.

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,60 @@ void SpotifySearchProvider::SearchFinishedSlot(
   PendingState state = it.value();
   queries_.erase(it);
 
+  QMap<std::string, std::string> album_dedup;
```
I think it would be better to add a comment to explain the logic we now apply here i.e. that we will check for duplicate in results.

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,60 @@ void SpotifySearchProvider::SearchFinishedSlot(
   for (int i = 0; i < response.album_size(); ++i) {
     const pb::spotify::Album& album = response.album(i);
 
+    QHash<QString, int> artist_count;
```
Maybe this could be in another helper function, to keep things clearer?

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,60 @@ void SpotifySearchProvider::SearchFinishedSlot(
     const pb::spotify::Album& album = response.album(i);
 
+    QHash<QString, int> artist_count;
+    QString artist, majority_artist;
```
I think artist can be defined later...

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,60 @@ void SpotifySearchProvider::SearchFinishedSlot(
+    // Choose an album artist
+    for (int j = 0; j < album.track_size(); ++j) {
+      for (int k = 0; k < album.track(j).artist_size(); ++k) {
+        artist = QStringFromStdString(album.track(j).artist(k));
```
... here, because you don't need it before, and don't reuse it outer the loop.
Usually, I prefer to keep the closest scope for object.
I don't think there is an overhead to do so, and even if there is one, it is probably very small.

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,60 @@ void SpotifySearchProvider::SearchFinishedSlot(
+        if (artist_count.contains(artist)) {
+          artist_count[artist]++;
+        } else {
+          artist_count[artist] = 1;
```
I don't think you need the else if/else, because IIRC trying to access an element that doesn't exist will create it with default value (0). That will make the code smaller.
That being said, maybe this behavior is a bit confusing and it's better to keep a the code as it is.

ArnaudBienner commented:
Thanks a lot for your work on this :)
The patch looks good (haven't tested it yet though). My comments are not mandatory things: just things which I think can make your patch even better :) The only thing I would like to have is at least small comment to briefly explain the logic we apply here.

cincodenada commented:
Thanks for the feedback!  I've added some clarifying comments (perhaps too much!), and moved `artist` down in the scope.  I left it in a separate commit so you can see the changes, but I'll squash things back down once it's ready to go.

ArnaudBienner commented on src/globalsearch/spotifysearchprovider.cpp:
```diff
@@ -90,30 +89,84 @@ void SpotifySearchProvider::SearchFinishedSlot(
+     * represented most frequently in the artist
+     *
+     * For most albums this will just be one artist,
+     * but this ensures we have a consistend album artist for
```
typo
s/consistend/consistent
